### PR TITLE
Add processing information to BAM header.

### PR DESCRIPTION
**Pull request by pontushojer**
I implemented a utility function that can add processing information to the BAM header. Currently this will add information about the program used (blr), the tool used (tagbam, clusterrmdup, etc.) and the command line .. command? (e.g. what was executed). Below is an example with the header from the final `*filt.bam` file. Note that the order of execution of each command is also maintained. What I would like to add later on is some kind of version number for the program. 

This addition will make it easy to track the exact processing performed on any BAM file, even should it be renamed.  

**Example. samtools view -H mapped.sorted.tag.mkdup.bcmerge.mol.filt.bam**
```  
@HD	VN:1.0	SO:coordinate
@SQ	SN:chr1mini	LN:100001
@PG	PN:bowtie2	ID:bowtie2	CL:"/Users/pontus.hojer/miniconda3/envs/blr/bin/bowtie2-align-s --wrapper basic-0 -p 4 --reorder --maxins 2000 -x ../testdata/chr1mini.fasta -1 trimmed.barcoded.1.fastq.gz -2 trimmed.barcoded.2.fastq.gz"	VN:2.3.5
@PG	PN:blr	ID:tagbam	CL:"/Users/pontus.hojer/miniconda3/envs/blr/bin/blr tagbam mapped.sorted.bam -o mapped.sorted.tag.bam -f sam"
@PG	ID:sambamba	CL:markdup -t 4 mapped.sorted.tag.bam mapped.sorted.tag.mkdup.bam	PP:tagbam	VN:0.6.6
@PG	PN:blr	ID:clusterrmdup	CL:"/Users/pontus.hojer/miniconda3/envs/blr/bin/blr clusterrmdup mapped.sorted.tag.mkdup.bam barcode-merges.csv -o mapped.sorted.tag.mkdup.bcmerge.bam -b BX"
@PG	PN:blr	ID:buildmolecules	CL:"/Users/pontus.hojer/miniconda3/envs/blr/bin/blr buildmolecules mapped.sorted.tag.mkdup.bcmerge.bam -o mapped.sorted.tag.mkdup.bcmerge.mol.bam --stats-files -m MI -n MN -b BX"
@PG	PN:blr	ID:filterclusters	CL:"/Users/pontus.hojer/miniconda3/envs/blr/bin/blr filterclusters mapped.sorted.tag.mkdup.bcmerge.mol.bam -m MI -n MN -b BX -M 260"
```

### CODE REVIEW

File: src/blr/utils.py [hunk 1]
```diff
@@ -76,3 +77,45 @@ def print_stats(summary, name=None, value_width=15, print_to=sys.stderr):
...
+    Create SAM header dict with new tool and command line argument information based on template file. Appends new PG
+    entry with tool name (ID), software name (PN), command line arguments (CL) to track the tools applied to the file.
+    Use in output SAM/BAM file as 'header' attribute.
+
+    :param file: string for SAM/BAM file
+    :param name: string '__name__' variable.
+    :param argv: list of arguments from sys.argv
+    :return: dict
+    """
+    def check_name(basename, prev_entries):
```
Comment (FrickTobias) [inline src/blr/utils.py]: Insert documentation.

File: src/blr/utils.py [hunk 2]
```diff
@@ -76,3 +77,45 @@ def print_stats(summary, name=None, value_width=15, print_to=sys.stderr):
...
     print("="*width, file=print_to)
+
+
+def create_header(file, name, argv):
+    """
+    Create SAM header dict with new tool and command line argument information based on template file. Appends new PG
+    entry with tool name (ID), software name (PN), command line arguments (CL) to track the tools applied to the file.
+    Use in output SAM/BAM file as 'header' attribute.
+
+    :param file: string for SAM/BAM file
```
Comment (FrickTobias) [inline src/blr/utils.py]: Is this a string describing the path to the BAM file?

File: src/blr/utils.py [hunk 3]
```diff
@@ -76,3 +77,45 @@ def print_stats(summary, name=None, value_width=15, print_to=sys.stderr):
...
+    :param name: string '__name__' variable.
```
Comment (FrickTobias) [inline src/blr/utils.py]: Do you mean that one should input `__name__` here?

File: src/blr/utils.py [hunk 4]
```diff
@@ -76,3 +77,45 @@ def print_stats(summary, name=None, value_width=15, print_to=sys.stderr):
...
+    :param argv: list of arguments from sys.argv
```
Comment (FrickTobias) [inline src/blr/utils.py]: This does not need to be passed to the script (sys.argv will be same if called inside this function).

File: src/blr/utils.py (hunk 1 above)
Comment (pontushojer) [reply src/blr/utils.py]: Will do!

File: src/blr/utils.py (hunk 2 above)
Comment (pontushojer) [reply src/blr/utils.py]: Yes, I will clarify in docs

File: src/blr/utils.py (hunk 3 above)
Comment (pontushojer) [reply src/blr/utils.py]: Yes!

File: src/blr/utils.py (hunk 4 above)
Comment (pontushojer) [reply src/blr/utils.py]: Ok, that might be true, I will check.

Comment (pontushojer) [reply src/blr/utils.py]: It is true, I will remove this as an argument and call in the util function.

**Pull request merged**